### ISSUE TITLE
Proposal: Move monitoring configuration closer to the service code, use TOML

### ISSUE THREAD

**Issue by slimsag** [sourcegraph/sourcegraph#12442]
### Problem

I have heard a lot of sentiment around the `/monitoring` generator, particularly from newcomers of it - that it's a bit scary to approach, things like:

- "I don't want to write code for this"
- "Why did we make our own complex monitoring generator instead of jsonnet"
- etc.

I think all of these stem from the fact that the monitoring generator uses Go code as its configuration format: it actually prohibits you from doing _a lot_ and the most complex parts of it (shared metrics) are [sternly advised against and regulated](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/monitoring/shared.go#L8-28).

### Proposal

Instead of:

- `/monitoring/frontend.go`
- `/monitoring/git_server.go`

I would like to have:

- `cmd/frontend/monitoring.toml`
- `cmd/gitserver/monitoring.toml`

### Benefits

This benefits us because:

- It lets us more strictly enforce what is and is not allowed in the monitoring generator, for example we can prohibit creating multiple `monitoring.toml` files below a single `cmd/service` directory with a linter easily - and nobody can add Go functions that do fancy things.
- A config file appears much simpler than a Go file.
- It living alongside the code of the service itself means it feels more like code the service author is responsible for, which aligns nicely with our team alerting/monitoring goals.

### Why TOML?

YAML multi-line support and indention is horrible, it will be much easier to write `PossibleSolutions` multi-line Markdown within TOML than YAML.

### Documentation

Documented the exact config format, restrictions etc., with a nice handbook entry will be very important here.

cc @bobheadxi @pecigonzalo

**Comment by uwedeportivo** [comment]
maybe have a look at dhall, in addition to the toml advantages it also gives you type safety and sophisticated diff detection (if your config is not materially changed, it knows it's a no-op). you can read it in just like all the others (https://github.com/philandstuff/dhall-golang)

**Comment by bobheadxi** [comment]
I'm actually a bit fan of the current Go-based approach - Go seems to be considered a core competency, and the completion/checking that comes with using Go to define monitoring configuration is really nice. IMO a TOML-based format just moves the same complexity elsewhere, and bring us a step closer to "why don't we just write Prometheus config"

**Comment by slimsag** [comment]
I strongly feel Dhall would not be a good choice here, that feels like raising the barrier even further for most on the team.

**Comment by slimsag** [comment]
I think your points / concerns here are completely fair @bobheadxi - and that does give me some pause admittedly. I also note @daxmc99 's sad face reaction on Slack - so I just want to clarify this is 100% a proposal not a made decision

Would love to hear more thoughts / feedback from others on this

**Comment by daxmc99** [comment]
I would like to echo @bobheadxi's sentiment here. I don't think TOML will reduce the complexity or make the code more palatable to newcomers. On some level, I think we need to decide if our monitoring code should be _mildly_ bespoke and not just a large amount of Prometheus + Grafana config files. 

I do believe in having the monitoring configuration closer to the source code they monitoring makes sense though. This spreads the monitoring + observability concern across teams. If a team does not like the Go approach I think they should be able to write just the raw config files and distribution can be responsible for getting them into the monitoring system.

**Comment by bobheadxi** [comment]
> If a team does not like the Go approach I think they should be able to write just the raw config files and distribution can be responsible for getting them into the monitoring system.

I agree regarding a format that helps monitoring feel a bit closer to home (maybe if we spread the current definitions into individual services?), but I'm hesitant to add more to our workload on this front by supporting a variety of formats - I think that effort might be better put into helping teams familiarize themselves with whatever format we finalize on

**Comment by daxmc99** [comment]
> I agree regarding a format that helps monitoring feel a bit closer to home (maybe if we spread the >current definitions into individual services?), but I'm hesitant to add more to our workload on this >front by supporting a variety of formats - I think that effort might be better put into helping teams > familiarize themselves with whatever format we finalize on

Ultimately, what [Grafana](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/tree/docker-images/grafana/config) & [Prometheus](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/tree/docker-images/prometheus/config) read are basic `json` and `yml` files (I assume). Could someone manually write one of these?

**Comment by uwedeportivo** [comment]
prometheus ones for sure. grafana ones are tedious

**Comment by pecigonzalo** [comment]
It would be great to define or link the original problem we are trying to fix. I believe there was some talk about it in relation to https://github.com/sourcegraph/about/pull/1221 but I cant find the content to link it directly, maybe it was talked about on a meeting.

The main benefit I see with the Go based generator is around generating dashboards for Grafana from the same code we use to define rules and all the wrapping we had to do because of `alert_count`. I think ideally services would define their dashboards and alerts automagically as part of their service definition/code like they define their metrics, although this could end up being even more complex.

The main downsides for me are around readability and easy to grok output from code because of the abstraction, this was made more complicated for me because part is its config is scattered between places (siteconfig, generator, static files, configmap).
Maybe this is not a problem of the generator itself and its just about making it simpler to onboard to by cleaning old configs, reducing the amount of places this is spread on and documenting what goes where and how.

As an example, without taking into account Grafana and wrapping, an alert rule without all the wrapping is quite simple to understand and even write.
```
groups:
- name: replacer
  rules:
  - alert: replacer_frontend_internal_api_error_responses
    expr: sum by(category) (increase(src_frontend_internal_request_duration_seconds_count{job="replacer",code!~"2.."}[5m])) > X
    labels:
      level: warning
      service_name: replacer
    annotations:
      summary: SOME summary
      description: SOME description
```

Or even "simpler" if we use generic alerts instead, as we will not have multiple copies of this same alert
```
groups:
- name: api_error_responses
  rules:
  - alert: frontend_internal_api_error_responses
    expr: sum by(category) (increase(src_frontend_internal_request_duration_seconds_count{code!~"2.."}[5m])) > X
    labels:
      level: warning
      service_name: "{{ $labels.job }}"
    annotations:
      summary: "SOME {{ $labels.foo }} summary"
      description: |
        SOME {{ $labels.bar }} description
```

Than the following Go definition, which I also need to go to each type and function to understand what it will do.
```
package main

func Replacer() *Container {
	return &Container{
		Name:        "replacer",
		Title:       "Replacer",
		Description: "Backend for find-and-replace operations.",
		Groups: []Group{
			{
				Title: "General",
				Rows: []Row{
					{
						sharedFrontendInternalAPIErrorResponses("replacer"),
					},
				},
			},
		},
	}
}
```

A rule with all the wrapping would be really ugly because we have to repeat the wrapper everywhere.
```
clamp_max(clamp_min(floor(
      max((((( 
THE_REAL_QUERY
) OR on() vector(0)) >= 0) OR on() vector(1))
      ), 0), 1) OR on() vector(1)
```

Therefore, I assume the generator is mainly there to help us resolve 3 things:
- Generating Grafana dashboards because they are cumbersome to write and read
- Generating wrapping due to `alert_count` and other current requirements (this could be changed/fixed)
- Ensuring alerts/rules conform to certain requirements

> I strongly feel Dhall would not be a good choice here, that feels like raising the barrier even further for most on the team.

I think this is a valid concern for Dhall right now, but if we use it for Kuberentes then it will be required anyway, otherwise the same concern applies there.

**Comment by slimsag** [comment]
I definitely do not want teams to end up having different ways of defining their alerts and dashboards, this sounds like the worst-case outcome to me:

- Some using the Go generator as a config file
- Some using the Go generator but doing fancy/complex things with it
- Some using Prometheus/Grafana config directly
- Some using Grafana WYSIWYG editor

> Ultimately, what Grafana & Prometheus read are basic json and yml files (I assume). Could someone manually write one of these?

I would rephrase htat question as "Would we *want* someone to do this?" and I believe the answer is no.

**Comment by slimsag** [comment]
Withdrawing this proposal for now.
